src/mpid/ch4/generic/am/mpidig_am_recv.h

```diff
@@ -158,6 +153,42 @@ MPL_STATIC_INLINE_PREFIX int MPIDIG_do_irecv(void *buf, MPI_Aint count, MPI_Data
             mpi_errno = MPIDIG_do_long_ack(unexp_req);
             MPIR_ERR_CHECK(mpi_errno);
             goto fn_exit;
+        } else if (MPIDIG_REQUEST(unexp_req, req->status) & MPIDIG_REQ_RTS) {
+            /* Matching receive is now posted, tell the netmod/shmmod */
+            /* first save the reference to the unexp buffer and data size in it */
+            void *first_seg = MPIDIG_REQUEST(unexp_req, buffer);
+            MPI_Aint first_seg_sz = 0;
+            /* the count in unexpected message is the total size of incoming data */
+            size_t in_data_sz = MPIDIG_REQUEST(unexp_req, count);
+            MPIDIG_get_recv_offset(unexp_req, &first_seg_sz);
+            /* Now it is safe to set request with user buffer address, datatype and count */
+            MPIR_Datatype_add_ref_if_not_builtin(datatype);
+            MPIDIG_REQUEST(unexp_req, datatype) = datatype;
+            MPIDIG_REQUEST(unexp_req, buffer) = (char *) buf;
+            MPIDIG_REQUEST(unexp_req, count) = count;
+            if (*request == NULL) {
+                /* Regular (non-enqueuing) path: MPIDIG is responsbile for allocating
+                 * a request. Here we simply return `unexp_req` */
+                *request = unexp_req;
+                /* Mark `match_req` as NULL so that we know nothing else to complete when
+                 * `unexp_req` finally completes. (See below) */
+                MPIDIG_REQUEST(unexp_req, req->rreq.match_req) = NULL;
+            } else {
+                /* Enqueuing path: CH4 already allocated a request.
+                 * Record the passed `*request` to `match_req` so that we can complete it
+                 * later when `unexp_req` completes.
+                 * See MPIDI_recv_target_cmpl_cb for actual completion handler. */
+                MPIDIG_REQUEST(unexp_req, req->rreq.match_req) = *request;
+            }
+            MPIDIG_REQUEST(unexp_req, req->status) &= ~MPIDIG_REQ_UNEXPECTED;
+            MPIDIG_recv_type_init(in_data_sz, unexp_req);
+            MPIDIG_recv_setup(unexp_req);
```

Comment:
Since we called `MPIDIG_get_recv_offset` earlier, I thought the `unexp_req` has already done `MPIDIG_recv_type_init`, why we do it again here? I guess that's because the initial buffer was small. In which case, we shouldn't hack the "async" routines, we should directly encode the first segment size. It is confusing as is.